src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,185 @@
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
+
+                var enabledPreviewFeatures = context.Options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.EnablePreviewFeatures, context.Compilation, context.CancellationToken);
+                if (string.Equals(enabledPreviewFeatures, "true", System.StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, IOperation?>();
+                context.RegisterOperationAction(context => BuildSymbolInformation(context, requiresPreviewFeaturesSymbols),
+                    OperationKind.Invocation,
+                    OperationKind.ObjectCreation,
+                    OperationKind.PropertyReference,
+                    OperationKind.FieldReference,
+                    OperationKind.DelegateCreation,
+                    OperationKind.EventReference
+                    );
+            });
+        }
+
+        private void BuildSymbolInformation(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            if (OperationUsesPreviewFeatures(context.Operation, requiresPreviewFeaturesSymbols))
+            {
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(Rule));
+            }
+        }
+
+        private bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            ISymbol? symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(operation, symbol, requiresPreviewFeaturesSymbols, true);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                _ => null,
+            };
+
+        private bool TryGetOrCachePreviewFeaturesAttributesOnSymbol(IOperation operation, ISymbol symbol, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols, bool checkParents)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out IOperation? existing))
```

Comment:
It looks like the dictionary is only used to check whether we encountered a symbol or not. Consider a HashSet for that.

Reply:
Yup, good catch. I missed doing that in the refactoring.

Reply:
I thought the cache will be used to determine if the symbol has the attribute or not, from this comment looks like it is only caching symbols that have the PreviewFeature attribute, which means for a non attributed symbol the cache is useless. I think the dictionary should be `<ISymbol, bool>` type then we could add all symbols into the cache with the value if it has the attribute